src/index.js

```diff
@@ -65,28 +77,88 @@ const decode = (dl, str) => {
   dl.identifier[segments.shift()] = segments.shift();
 
   // /x/y until query
-  while(segments.length) {
+  while (segments.length) {
     dl.keyQualifiers[segments.shift()] = segments.shift();
   }
 
   // ?x=y...
   if (str.includes('?')) {
-    str.substring(str.indexOf('?') + 1)
+    str
+      .substring(str.indexOf('?') + 1)
       .split('&')
-      .forEach((pair) => {
+      .forEach(pair => {
         const [key, value] = pair.split('=');
         dl.attributes[key] = value;
       });
   }
 };
 
+/**
+ * Extract from the grammar rules the order of all the possible key qualifiers and return a map with a weight for each parameter
+ *
+ * @returns {Map} A map that contains all the possible key qualifiers and their weights. Example : { '10': 2, '21': 1, '22': 3, cpv: 3, lot: 2, ser: 1 }
+ */
+const giveTheKeyQualifiersAWeight = () => {
+  const rule = findTheRule(Rules.gtinPath);
+
+  if (!rule) throw new Error(`The rule ${rule} wasn't find in the grammar file`);
+
+  // if my rule is : 'gtin-path  = gtin-comp  [cpv-comp] [lot-comp] [ser-comp]'
+  // In my parameters array, I'll have : [ 'cpv-comp', 'lot-comp', 'ser-comp' ]
+  const parameters = rule
+    .substring(rule.indexOf('=') + 1)
+    .split(' ')
+    .map(item => item.replace('/ /g', ''))
+    .filter(item => item.startsWith('['))
+    .map(item => item.replace('[', '').replace(']', ''));
+
+  // I retrieve the code rules of each parameter
+  // [ 'cpv-code  = "22" / %s"cpv"   ; Consumer Product Variant',
+  //   'lot-code  = "10" / %s"lot"   ; Batch/Lot identifier',
+  //   'ser-code  = "21" / %s"ser"   ; GTIN Serial Number' ]
+  const parametersCodeRules = parameters.map(param => findTheRule(param.replace('comp', 'code')));
+
+  // I retrieve the code of each parameter : [ '22', '10', '21' ]
+  const parametersCodes = parametersCodeRules.map(singleRule => singleRule.match(/(\d+)/)[0]);
+
+  const keyQualifiersCodeWeight = parametersCodes.map((elem, index) => {
+    const str = elem.toString();
+    const map = {};
+    map[str] = parameters.length - index;
+    return map;
+  });
+
+  const keyQualifiersNameWeight = parameters.map((elem, index) => {
+    const str = elem.toString().replace('-comp', '');
+    const map = {};
+    map[str] = parameters.length - index;
+    return map;
+  });
+
+  const keyQualifiersWeight = new Map();
+
+  // I add the two maps to the keyQualifiersWeight Map
+  keyQualifiersNameWeight.forEach(elem => {
+    Object.keys(elem).forEach(item => {
+      keyQualifiersWeight[item] = elem[item];
+    });
+  });
+  keyQualifiersCodeWeight.forEach(elem => {
+    Object.keys(elem).forEach(item => {
```

Comment:
`Object.entries(([name, value]) => { ... })` could be simpler and a bit more descriptive.